LLM/sub_model/query_index_loader.py

```
@@ -0,0 +1,117 @@
+
+def load_embeddings(path: Path) -> np.ndarray:
+    embeddings = np.load(path, mmap_mode="r").astype(np.float32)
+    row_norms = np.linalg.norm(embeddings, axis=1, keepdims=True)
+    row_norms[row_norms == 0] = 1.0
+    return embeddings / row_norms
+
+
+def load_bm25(path: Path, tok_path: Path, search_df: pd.DataFrame, tokenizer) -> BM25Okapi:
+    if path.exists():
+        with open(path, "rb") as f:
+            return pickle.load(f)
```

Ruff (0.15.15)

[error] `pickle` and modules that wrap it can be unsafe when used to deserialize untrusted data, possible security issue

(S301)